guides/source/active_model_basics.md

```diff
@@ -122,151 +453,439 @@ class Person
+  `define_attribute_methods`.
+3. Call `[attr_name]_will_change!` before each change to the tracked attribute.
+4. Call `changes_applied` after the changes are persisted.
+5. Call `clear_changes_information` when you want to reset the changes
+   information.
+6. Call `restore_attributes` when you want to restore previous data.
+
+You can then use the methods provided by `ActiveModel::Dirty` to query the
+object for its list of all changed attributes, the original values of the
+changed attributes, and the changes made to the attributes.
+
+Let's consider a Person class with attributes `first_name` and `last_name` and
+determine how we can use `ActiveModel::Dirty` to track changes to these
+attributes.
 
 ```ruby
 class Person
   include ActiveModel::Dirty
+
+  attr_reader :first_name, :last_name
   define_attribute_methods :first_name, :last_name
 
-  def first_name
-    @first_name
+  def initialize
+    @first_name = nil
+    @last_name = nil
   end
 
   def first_name=(value)
-    first_name_will_change!
+    first_name_will_change! unless value == @first_name
     @first_name = value
   end
 
-  def last_name
-    @last_name
-  end
-
   def last_name=(value)
-    last_name_will_change!
+    last_name_will_change! unless value == @last_name
     @last_name = value
   end
 
   def save
-    # do save work...
+    # Persist data - clears dirty data and moves `changes` to `previously_changed`.
     changes_applied
   end
+
+  def reload!
+    # Clears all dirty data: current changes and previous changes.
+    clear_changes_information
+  end
+
+  def rollback!
+   # Restores all previous data of the provided attributes.
+    restore_attributes
+  end
 end
 ```
 
 #### Querying an Object Directly for its List of All Changed Attributes
 
 ```irb
 irb> person = Person.new
+
+# A newly instantiated `Person` object is unchanged:
 irb> person.changed?
 => false
 
-irb> person.first_name = "First Name"
+irb> person.first_name = "Jane Doe"
 irb> person.first_name
-=> "First Name"
+=> "Jane Doe"
+```
 
-# Returns true if any of the attributes have unsaved changes.
+**`changed?`** returns true if any of the attributes have unsaved changes,
+`false` otherwise.
+
+```irb
 irb> person.changed?
 => true
+```
+
+**`changed`** returns an array with the name of the attributes containing
+unsaved changes.
 
-# Returns a list of attributes that have changed before saving.
+```irb
 irb> person.changed
 => ["first_name"]
+```
+
+**`changed_attributes`** returns a hash of the attributes with unsaved changes
+indicating their original values like `attr => original value`.
 
-# Returns a Hash of the attributes that have changed with their original values.
+```irb
 irb> person.changed_attributes
-=> {"first_name"=>nil}
+=> {"first_name" => nil}
+```
+
+**`changes`** returns a hash of changes, with the attribute names as the keys,
+and the values as an array of the original and new values like `attr =>
+[original value, new value]`.
```

Comment:
It's not a problem if we go over the line length here and there, fine to do it on these.

(it might just make "auto wrapping" not work in the whole file, but it's seems unlikely that'll be done again any time soon anyway.)